src/ui/public/visualize/visualize.js

```diff
@@ -159,38 +153,30 @@ uiModules
             if ($scope.vis.type.requiresSearch && ['query', 'filters'].includes(keys[0])) {
               $scope.fetch();
             }
-            if (keys[0] === 'uiState') {
-            // uiState can be changed by other visualizations on dashboard. this makes sure this fires only if
-            // current visualizations uiState changed.
-              if (!oldUiState || oldUiState !== JSON.stringify($scope.uiState.toJSON())) {
-                oldUiState = JSON.stringify($scope.uiState.toJSON());
-                $scope.fetch();
-              }
-            }
           });
 
           $scope.$on('$destroy', () => {
             stateMonitor.destroy();
           });
-        } else {
-          const handleUiStateChange = () => { $scope.$broadcast('render'); };
```

Comment:
If there wasn't an `appState` available we would already watch on the `uiState` directly. This is now replaced by the general watch on `uiState` further down this file (no matter if `appState` existed or not).

Also we just triggered a render when the `uiState` changes, and not a new `fetch`, making the assumption, that nothing in the `uiState` would require new data to be loaded.

This assumption was rather dangerous, and we now always trigger a fetch, when the `uiState` changes. If nothing in the `uiState` changed, that would require new data, we rely on the actual request handler to detect that, and serve a cached answer instead.